File: plugins/capacitor-plugin-simple-encryption/android/src/main/java/cash/selene/plugins/simpleencryption/SimpleEncryptionPlugin.kt
package cash.selene.plugins.simpleencryption

import android.util.Base64
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import javax.crypto.Cipher

@CapacitorPlugin(name = "SimpleEncryption")
class SimpleEncryptionPlugin : Plugin() {
    private lateinit var keyManager: KeyManager

    override fun load() {
        keyManager = KeyManager(context)
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        val pin = call.getString("pin")

        try {
            keyManager.initialize(pin)
            val result = JSObject()
            result.put("isNative", true)
            result.put("hasPinConfigured", keyManager.hasPinConfigured())
            result.put("isReady", keyManager.isReady)
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Initialization failed", e)
        }
    }

    @PluginMethod
    fun hasPinConfigured(call: PluginCall) {
        val result = JSObject()
        result.put("value", keyManager.hasPinConfigured())
        call.resolve(result)
    }

    @PluginMethod
    fun setPin(call: PluginCall) {
        val newPin = call.getString("newPin")
        if (newPin == null) {
            call.reject("newPin is required")
            return
        }

        try {
            keyManager.setPin(newPin)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to set PIN", e)
        }
    }

    @PluginMethod
    fun removePin(call: PluginCall) {
        try {
            keyManager.removePin()
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to remove PIN", e)
        }
    }

    @PluginMethod
    fun verifyPin(call: PluginCall) {
        val pin = call.getString("pin")
        if (pin == null) {
            call.reject("pin is required")
            return
        }

        try {
            val isValid = keyManager.verifyPin(pin)
            val result = JSObject()
            result.put("isValid", isValid)
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to verify PIN", e)
        }
    }

    @PluginMethod
    fun encrypt(call: PluginCall) {
        val plainString = call.getString("data")
        if (plainString == null) {
            call.reject("data is required")
            return
        }

        var data: ByteArray? = null
        try {
            data = plainString.toByteArray(Charsets.UTF_8)
            val encrypted = keyManager.encrypt(data)
            val result = JSObject()
            result.put("data", Base64.encodeToString(encrypted, Base64.NO_WRAP))
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Encryption failed", e)
        } finally {
            data?.fill(0)
        }
    }

    @PluginMethod
    fun decrypt(call: PluginCall) {
        val encryptedBase64 = call.getString("data")
        if (encryptedBase64 == null) {
            call.reject("data is required")
            return
        }

        var data: ByteArray? = null
        var decrypted: ByteArray? = null
        try {
            data = Base64.decode(encryptedBase64, Base64.NO_WRAP)
            decrypted = keyManager.decrypt(data)
            val result = JSObject()
            result.put("data", String(decrypted, Charsets.UTF_8))
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Decryption failed", e)
        } finally {
            data?.fill(0)
            decrypted?.fill(0)
        }
    }

    // ----Key Storage Settings

    @PluginMethod
    fun setKeyStorageSettings(call: PluginCall) {
        val deviceOnly = call.getBoolean("deviceOnly") ?: false

        try {
            keyManager.setDeviceOnly(deviceOnly)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to set key storage settings", e)
        }
    }

    // ----Key Backup/Restore

    @PluginMethod
    fun exportKeyBackup(call: PluginCall) {
        val password = call.getString("password")
        if (password == null) {
            call.reject("password is required")
            return
        }

        try {
            val backupData = keyManager.exportKeyBackup(password)
            val result = JSObject()
            result.put("data", Base64.encodeToString(backupData, Base64.NO_WRAP))
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to export key backup", e)
        }
    }

    @PluginMethod
    fun importKeyBackup(call: PluginCall) {
        val dataBase64 = call.getString("data")
        if (dataBase64 == null) {
            call.reject("data is required")
            return
        }

        val password = call.getString("password")
        if (password == null) {
            call.reject("password is required")
            return
        }

        try {
            val data = Base64.decode(dataBase64, Base64.NO_WRAP)
            keyManager.importKeyBackup(data, password)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to import key backup", e)
        }
    }

    // ----Re-encryption Support

    @PluginMethod
    fun exportCurrentKey(call: PluginCall) {
        var key: ByteArray? = null
        try {
            key = keyManager.exportCurrentKey()
            val result = JSObject()
            result.put("key", Base64.encodeToString(key, Base64.NO_WRAP))
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to export key", e)
        } finally {
            key?.fill(0)
        }
    }

    @PluginMethod
    fun decryptWithExplicitKey(call: PluginCall) {
        val encryptedBase64 = call.getString("data")
        if (encryptedBase64 == null) {
            call.reject("data is required")
            return
        }

        val keyBase64 = call.getString("key")
        if (keyBase64 == null) {
            call.reject("key is required")
            return
        }

        var key: ByteArray? = null
        var decrypted: ByteArray? = null
        try {
            key = Base64.decode(keyBase64, Base64.NO_WRAP)
            val data = Base64.decode(encryptedBase64, Base64.NO_WRAP)
            decrypted = keyManager.decryptWithKey(data, key)
            val result = JSObject()
            result.put("data", String(decrypted, Charsets.UTF_8))
            call.resolve(result)
        } catch (e: Exception) {
            call.reject(e.message ?: "Decryption failed", e)
        } finally {
            key?.fill(0)
            decrypted?.fill(0)
        }
    }

    @PluginMethod
    fun loadKeyIntoMemory(call: PluginCall) {
        val keyBase64 = call.getString("key")
        if (keyBase64 == null) {
            call.reject("key is required")
            return
        }

        try {
            val key = Base64.decode(keyBase64, Base64.NO_WRAP)
            keyManager.loadKeyIntoMemory(key)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to load key into memory", e)
        }
    }

    @PluginMethod
    fun replaceKey(call: PluginCall) {
        val keyBase64 = call.getString("key")
        if (keyBase64 == null) {
            call.reject("key is required")
            return
        }

        try {
            val key = Base64.decode(keyBase64, Base64.NO_WRAP)
            keyManager.replaceKey(key)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to replace key", e)
        }
    }

    // ----Biometric Key Storage

    @PluginMethod
    fun storeBiometricKey(call: PluginCall) {
        val keyBase64 = call.getString("key")
        if (keyBase64 == null) {
            call.reject("key is required")
            return
        }

        val activity = activity as? FragmentActivity
        if (activity == null) {
            call.reject("Activity not available for biometric prompt")
            return
        }

        val title = call.getString("title") ?: "Selene Wallet"
        val reason = call.getString("reason") ?: "Authenticate to enable biometric unlock"

        try {
            val key = Base64.decode(keyBase64, Base64.NO_WRAP)
            val cipher = keyManager.getBiometricEncryptCipher()
            val executor = ContextCompat.getMainExecutor(context)

            val callback = object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    try {
                        val authenticatedCipher = result.cryptoObject?.cipher
                            ?: throw EncryptionException.CryptoError("No cipher from biometric result")
                        keyManager.storeBiometricKey(key, authenticatedCipher)
                        call.resolve()
                    } catch (e: Exception) {
                        call.reject(e.message ?: "Failed to store biometric key", e)
                    } finally {
                        key.fill(0)
                    }
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    key.fill(0)
                    call.reject("Biometric authentication failed: $errString")
                }

                override fun onAuthenticationFailed() {
                    // Called on individual attempt failure, not final - don't reject yet
                }
            }

            activity.runOnUiThread {
                val prompt = BiometricPrompt(activity, executor, callback)
                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle(title)
                    .setSubtitle(reason)
                    .setNegativeButtonText(context.getString(R.string.biometric_prompt_cancel))
                    .build()
                prompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
            }
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to store biometric key", e)
        }
    }

    @PluginMethod
    fun loadBiometricKey(call: PluginCall) {
        val activity = activity as? FragmentActivity
        if (activity == null) {
            call.reject("Activity not available for biometric prompt")
            return
        }

        val title = call.getString("title") ?: "Authorize"
        val reason = call.getString("reason") ?: ""

        try {
            val cipher = keyManager.getBiometricDecryptCipher()
            val executor = ContextCompat.getMainExecutor(context)

            val callback = object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    var key: ByteArray? = null
                    try {
                        val authenticatedCipher = result.cryptoObject?.cipher
                            ?: throw EncryptionException.CryptoError("No cipher from biometric result")
                        key = keyManager.decryptBiometricKey(authenticatedCipher)
                        val jsResult = JSObject()
                        jsResult.put("key", Base64.encodeToString(key, Base64.NO_WRAP))
                        call.resolve(jsResult)
                    } catch (e: Exception) {
                        call.reject(e.message ?: "Failed to decrypt biometric key", e)
                    } finally {
                        key?.fill(0)
                    }
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    call.reject("Biometric authentication failed: $errString")
                }

                override fun onAuthenticationFailed() {
                    // Called on individual attempt failure, not final - don't reject yet
                }
            }

            activity.runOnUiThread {
                val prompt = BiometricPrompt(activity, executor, callback)
                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle(title)
                    .setSubtitle(reason)
                    .setNegativeButtonText(context.getString(R.string.biometric_prompt_cancel))
                    .build()
                prompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
            }
        } catch (e: Exception) {
            call.reject(e.message ?: "Failed to load biometric key", e)
        }
    }

    @PluginMethod
    fun hasBiometricKey(call: PluginCall) {
        val result = JSObject()
        result.put("value", keyManager.hasBiometricKey())
        call.resolve(result)
    }

    @PluginMethod
    fun removeBiometricKey(call: PluginCall) {
        keyManager.deleteBiometricKey()
        call.resolve()
    }

    // ----Biometric Availability & Verification

    @PluginMethod
    fun isBiometricAvailable(call: PluginCall) {
        val canAuthenticate = BiometricManager.from(context)
            .canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)
        val result = JSObject()
        result.put("value", canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS)
        call.resolve(result)
    }

    @PluginMethod
    fun verifyBiometric(call: PluginCall) {
        val activity = activity as? FragmentActivity
        if (activity == null) {
            call.reject("Activity not available for biometric prompt")
            return
        }

        val title = call.getString("title") ?: "Authorize"
        val reason = call.getString("reason") ?: ""

        val executor = ContextCompat.getMainExecutor(context)

        val callback = object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                call.resolve()
            }

            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                call.reject("Biometric authentication failed: $errString")
            }

            override fun onAuthenticationFailed() {
                // Called on individual attempt failure, not final - don't reject yet
            }
        }

        activity.runOnUiThread {
            val prompt = BiometricPrompt(activity, executor, callback)
            val promptInfo = BiometricPrompt.PromptInfo.Builder()
                .setTitle(title)
                .setSubtitle(reason)
                .setNegativeButtonText(context.getString(R.string.biometric_prompt_cancel))
                .build()
            prompt.authenticate(promptInfo)
        }
    }

    // ----Key Clearing

    @PluginMethod
    fun clearKeyFromMemory(call: PluginCall) {
        keyManager.clearKey()
        call.resolve()
    }

    // ----Nuclear Reset

    @PluginMethod
    fun resetAll(call: PluginCall) {
        keyManager.resetAll()
        call.resolve()
    }

    // ----App Settings (no-op on Android — no per-app biometric permissions)

    @PluginMethod
    fun openAppSettings(call: PluginCall) {
        call.resolve()
    }
}

File: plugins/capacitor-plugin-simple-encryption/ios/Sources/SimpleEncryptionPlugin/SimpleEncryptionPlugin.swift
import Foundation
import Capacitor
import CryptoKit
import LocalAuthentication

@objc(SimpleEncryptionPlugin)
public class SimpleEncryptionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SimpleEncryptionPlugin"
    public let jsName = "SimpleEncryption"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hasPinConfigured", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setPin", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removePin", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "verifyPin", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "encrypt", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "decrypt", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setKeyStorageSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "exportKeyBackup", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "importKeyBackup", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "exportCurrentKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "decryptWithExplicitKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "replaceKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadKeyIntoMemory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storeBiometricKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadBiometricKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hasBiometricKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeBiometricKey", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isBiometricAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "verifyBiometric", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearKeyFromMemory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resetAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAppSettings", returnType: CAPPluginReturnPromise)
    ]

    private let keyManager = KeyManager()

    // MARK: - Plugin Methods

    @objc func initialize(_ call: CAPPluginCall) {
        let pin = call.getString("pin")

        do {
            try keyManager.initialize(pin: pin)

            call.resolve([
                "isNative": true,
                "hasPinConfigured": keyManager.hasPinConfigured(),
                "isReady": keyManager.isReady
            ])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func hasPinConfigured(_ call: CAPPluginCall) {
        call.resolve(["value": keyManager.hasPinConfigured()])
    }

    @objc func setPin(_ call: CAPPluginCall) {
        guard let newPin = call.getString("newPin") else {
            call.reject("newPin is required")
            return
        }

        do {
            try keyManager.setPin(newPin: newPin)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func removePin(_ call: CAPPluginCall) {
        do {
            try keyManager.removePin()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func verifyPin(_ call: CAPPluginCall) {
        guard let pin = call.getString("pin") else {
            call.reject("pin is required")
            return
        }

        do {
            let isValid = try keyManager.verifyPin(pin)
            call.resolve(["isValid": isValid])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func encrypt(_ call: CAPPluginCall) {
        guard let plainString = call.getString("data") else {
            call.reject("data is required")
            return
        }

        guard let data = plainString.data(using: .utf8) else {
            call.reject("Failed to encode string as UTF-8")
            return
        }

        do {
            let encrypted = try keyManager.encrypt(data: data)
            call.resolve(["data": encrypted.base64EncodedString()])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func decrypt(_ call: CAPPluginCall) {
        guard let encryptedBase64 = call.getString("data") else {
            call.reject("data is required")
            return
        }

        guard let data = Data(base64Encoded: encryptedBase64) else {
            call.reject("Invalid base64 data")
            return
        }

        do {
            let decrypted = try keyManager.decrypt(data: data)
            guard let plainString = String(data: decrypted, encoding: .utf8) else {
                call.reject("Failed to decode decrypted data as UTF-8")
                return
            }
            call.resolve(["data": plainString])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    // MARK: - Key Storage Settings

    @objc func setKeyStorageSettings(_ call: CAPPluginCall) {
        let deviceOnly = call.getBool("deviceOnly") ?? false

        do {
            try keyManager.setDeviceOnly(deviceOnly)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    // MARK: - Key Backup/Restore

    @objc func exportKeyBackup(_ call: CAPPluginCall) {
        guard let password = call.getString("password") else {
            call.reject("password is required")
            return
        }

        do {
            let backupData = try keyManager.exportKeyBackup(password: password)
            call.resolve(["data": backupData.base64EncodedString()])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func importKeyBackup(_ call: CAPPluginCall) {
        guard let dataBase64 = call.getString("data") else {
            call.reject("data is required")
            return
        }

        guard let password = call.getString("password") else {
            call.reject("password is required")
            return
        }

        guard let data = Data(base64Encoded: dataBase64) else {
            call.reject("Invalid base64 data")
            return
        }

        do {
            try keyManager.importKeyBackup(data: data, password: password)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    // MARK: - Re-encryption Support

    @objc func exportCurrentKey(_ call: CAPPluginCall) {
        do {
            let key = try keyManager.exportCurrentKey()
            call.resolve(["key": key.base64EncodedString()])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func decryptWithExplicitKey(_ call: CAPPluginCall) {
        guard let encryptedBase64 = call.getString("data") else {
            call.reject("data is required")
            return
        }

        guard let keyBase64 = call.getString("key") else {
            call.reject("key is required")
            return
        }

        guard var key = Data(base64Encoded: keyBase64) else {
            call.reject("Invalid base64 key")
            return
        }
        defer { keyManager.zeroData(&key) }

        guard let data = Data(base64Encoded: encryptedBase64) else {
            call.reject("Invalid base64 data")
            return
        }

        do {
            let decrypted = try keyManager.decryptAESGCM(data: data, key: key)
            guard let plainString = String(data: decrypted, encoding: .utf8) else {
                call.reject("Failed to decode decrypted data as UTF-8")
                return
            }
            call.resolve(["data": plainString])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func loadKeyIntoMemory(_ call: CAPPluginCall) {
        guard let keyBase64 = call.getString("key") else {
            call.reject("key is required")
            return
        }

        guard let key = Data(base64Encoded: keyBase64) else {
            call.reject("Invalid base64 key")
            return
        }

        do {
            try keyManager.loadKeyIntoMemory(key)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    @objc func replaceKey(_ call: CAPPluginCall) {
        guard let keyBase64 = call.getString("key") else {
            call.reject("key is required")
            return
        }

        guard let key = Data(base64Encoded: keyBase64) else {
            call.reject("Invalid base64 key")
            return
        }

        do {
            try keyManager.replaceKey(key)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }

    // MARK: - Biometric Key Storage

    @objc func storeBiometricKey(_ call: CAPPluginCall) {
        guard let keyBase64 = call.getString("key") else {
            call.reject("key is required")
            return
        }

        guard var key = Data(base64Encoded: keyBase64) else {
            call.reject("Invalid base64 key")
            return
        }

        let title = call.getString("title") ?? "Selene Wallet"
        let reason = call.getString("reason") ?? "Authenticate to enable biometric unlock"

        // Verify biometric before storing — matches Android BiometricPrompt behavior
        DispatchQueue.global(qos: .userInitiated).async {
            let context = LAContext()
            context.localizedFallbackTitle = ""
            context.localizedCancelTitle = title

            var authError: NSError?
            guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &authError) else {
                self.keyManager.zeroData(&key)
                call.reject("Biometric not available: \(authError?.localizedDescription ?? "unknown")")
                return
            }

            context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, error in
                defer { self.keyManager.zeroData(&key) }

                if !success {
                    call.reject("Biometric authentication failed: \(error?.localizedDescription ?? "unknown")")
                    return
                }

                do {
                    try self.keyManager.storeBiometricKey(key)
                    call.resolve()
                } catch {
                    call.reject(error.localizedDescription, nil, error)
                }
            }
        }
    }

    @objc func loadBiometricKey(_ call: CAPPluginCall) {
        let reason = call.getString("reason")

        // Run on background thread - biometric prompt blocks
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                var key = try self.keyManager.loadBiometricKey(reason: reason)
                let encoded = key.base64EncodedString()
                self.keyManager.zeroData(&key)
                call.resolve(["key": encoded])
            } catch {
                call.reject(error.localizedDescription, nil, error)
            }
        }
    }

    @objc func hasBiometricKey(_ call: CAPPluginCall) {
        call.resolve(["value": keyManager.hasBiometricKey()])
    }

    @objc func removeBiometricKey(_ call: CAPPluginCall) {
        keyManager.deleteBiometricKey()
        call.resolve()
    }

    @objc func isBiometricAvailable(_ call: CAPPluginCall) {
        let context = LAContext()
        let isAvailable = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: nil)
        call.resolve(["value": isAvailable])
    }

    @objc func verifyBiometric(_ call: CAPPluginCall) {
        var reason = call.getString("reason") ?? "Authorize"
        if reason.isEmpty { reason = "Authorize" }

        // Run on background thread - biometric prompt blocks
        DispatchQueue.global(qos: .userInitiated).async {
            let context = LAContext()
            context.localizedFallbackTitle = "" // Hide device passcode fallback

            var authError: NSError?
            guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &authError) else {
                call.reject("Biometric authentication not available: \(authError?.localizedDescription ?? "unknown")")
                return
            }

            context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, error in
                if success {
                    call.resolve()
                } else {
                    call.reject("Biometric authentication failed: \(error?.localizedDescription ?? "unknown")")
                }
            }
        }
    }

    // MARK: - Key Clearing

    @objc func clearKeyFromMemory(_ call: CAPPluginCall) {
        keyManager.clearKey()
        call.resolve()
    }

    // MARK: - Nuclear Reset

    @objc func resetAll(_ call: CAPPluginCall) {
        keyManager.resetAll()
        call.resolve()
    }

    // MARK: - App Settings

    @objc func openAppSettings(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString) else {
                call.reject("Cannot open settings URL")
                return
            }
            UIApplication.shared.open(url, options: [:]) { success in
                if success {
                    call.resolve()
                } else {
                    call.reject("Failed to open settings")
                }
            }
        }
    }

}
